sdk/lib/opentelemetry/sdk/trace/samplers/probability_sampler.rb

```diff
@@ -26,51 +30,51 @@ def initialize(probability, hints:, ignore_parent:, apply_to_root_spans:, apply_
+            # sampled_flag = @use_parent_sampled_flag && parent_context&.trace_flags&.sampled?
+            # sampled_flag ||= hint == HINT_RECORD_AND_PROPAGATE
+            # sampled_flag ||= (parent_context.nil? || @apply_to_all_spans || (@apply_to_remote_parent && parent_context.remote?)) && probably(trace_id)
+            #
+            # More factored (readable?) form
+            sampled_flag = sample(hint, trace_id, parent_context)
+            record_events = hint == HINT_RECORD || sampled_flag
 
-          # Take the hint if one is provided and we're not ignoring it.
-          def take_hint(hint)
-            @result_from_hint[hint] if hint
-          end
-
-          # If the parent is sampled and we're not ignoring it keep the sampling decision.
-          def use_parent_sampling(parent_context)
-            RECORD_AND_PROPAGATE if !@ignore_parent && parent_context&.trace_flags&.sampled?
+            if sampled_flag && record_events
+              RECORD_AND_PROPAGATE
+            elsif record_events
+              RECORD
+            else
+              NOT_RECORD
+            end
           end
 
-          # If any link is sampled keep the sampling decision.
-          def use_link_sampling(links)
-            RECORD_AND_PROPAGATE if links&.any? { |link| link.context.trace_flags.sampled? }
-          end
+          private
 
-          def maybe_dont_apply(parent_context)
-            dont_apply_to_root_span(parent_context) ||
-              dont_apply_to_remote_parent(parent_context) ||
-              dont_apply_to_local_child(parent_context)
-          end
+          # Explicitly ignore these parameters.
+          def ignore(_links, _name, _kind, _attributes); end
 
-          def dont_apply_to_root_span(parent_context)
-            NOT_RECORD if !@apply_to_root_spans && parent_context.nil?
+          def sample(hint, trace_id, parent_context)
+            if parent_context.nil?
+              hint == HINT_RECORD_AND_PROPAGATE || probably(trace_id)
+            else
+              parent_sampled_flag(parent_context) || hint == HINT_RECORD_AND_PROPAGATE || probably_for_child(parent_context, trace_id)
+            end
           end
 
-          def dont_apply_to_remote_parent(parent_context)
-            NOT_RECORD if !@apply_to_remote_parent && parent_context&.remote?
+          def parent_sampled_flag(parent_context)
+            @use_parent_sampled_flag && parent_context.trace_flags.sampled?
           end
 
-          def dont_apply_to_local_child(parent_context)
-            NOT_RECORD if !@apply_to_all_spans && parent_context && !parent_context.remote?
+          def probably_for_child(parent_context, trace_id)
+            (@apply_to_all_spans || (@apply_to_remote_parent && parent_context.remote?)) && probably(trace_id)
           end
 
-          def use_probability_sampling(trace_id)
-            RECORD_AND_PROPAGATE if @probability == 1.0 || trace_id[16, 16] < @id_upper_bound
+          def probably(trace_id)
```

Comment:
I'm wondering if we can have slightly more descriptive name for this (and probably `probably_for_child`). I'm thinking something along the lines of `sample_trace_id?`.